File: src/pkg/addon/wasm.ts
import type {Binding, Bindings} from '../addon';
import {LAYOUT} from '../data';

export function toBindings<T extends boolean>(w: WebAssembly.Exports): Bindings<T> {
  return {
    options: {
      showdown: w.SHOWDOWN.valueOf(),
      log: w.LOG.valueOf(),
      chance: w.CHANCE.valueOf(),
      calc: w.CALC.valueOf(),
    },
    bindings: [toBinding(1, w)],
  };
}

function toBinding(gen: number, w: WebAssembly.Exports): Binding {
  const prefix = `GEN${gen}`;
  const size = LAYOUT[gen - 1].sizes.Battle;
  const update = w[`${prefix}_update`] as CallableFunction;
  const choices = w[`${prefix}_choices`] as CallableFunction;
  const memory = new Uint8Array((w.memory as WebAssembly.Memory).buffer);

  return {
    CHOICES_SIZE: w[`${prefix}_CHOICES_SIZE`].valueOf(),
    LOGS_SIZE: w[`${prefix}_LOGS_SIZE`].valueOf(),
    update(battle: ArrayBuffer, c1: number, c2: number, log: ArrayBuffer | undefined): number {
      memory.set(battle as any);
      if (log) {
        memory.set(log as any, size);
        return update(0, c1, c2, size);
      } else {
        return update(0, c1, c2, 0);
      }
    },
    choices(battle: ArrayBuffer, player: number, request: number, options: Uint8Array): number {
      memory.set(battle as any);
      const n = choices(0, player, request, size);
      for (let i = 0; i < n; i++) options[i] = memory[size + i];
      return n;
    },
  };
}

File: src/lib/wasm.zig
const std = @import("std");
const wasm = @import("bindings/wasm.zig");

export const SHOWDOWN = wasm.options.showdown;
export const LOG = wasm.options.log;
export const CHANCE = wasm.options.chance;
export const CALC = wasm.options.calc;

export const GEN1_CHOICES_SIZE = wasm.gen(1).CHOICES_SIZE;
export const GEN1_LOGS_SIZE = wasm.gen(1).LOGS_SIZE;

const exportable = @hasDecl(std.zig, "Zir") and !@hasDecl(std.zig.Zir.Inst, "export_value");

usingnamespace if (exportable) struct {
    export const GEN1_update = wasm.gen(1).update;
    export const GEN1_choices = wasm.gen(1).choices;
} else wasm.exports();
